src/backend/dx12/src/native.rs

```diff
@@ -401,21 +404,107 @@ pub struct DescriptorHeapSlice {
-            Some(DualHandle {
-                cpu: d3d12::D3D12_CPU_DESCRIPTOR_HANDLE { ptr: self.start.cpu.ptr + (self.handle_size * index) as usize },
-                gpu: d3d12::D3D12_GPU_DESCRIPTOR_HANDLE { ptr: self.start.gpu.ptr + (self.handle_size * index) as u64 },
-            })
-        } else {
-            None
+        let mut best_range_tuple: Option<(usize, Range<u64>)> = None;
+        for range in self.empty_ranges.iter().cloned().enumerate() {
+            if range.1.end - range.1.start < count {
+                continue;
+            }
+            match best_range_tuple {
+                Some((ref mut index, ref mut best_range)) => {
+                    // We want to find the best fit for this allocation in order to reduce 
+                    // memory fragmentation, so we'll look for it here.
+                    if range.1.end - range.1.start < best_range.end - best_range.start {
+                        *index = range.0;
+                        *best_range = range.1.clone();
+                        // Found a perfect fit, so stop looking.
+                        if range.1.end - range.1.start == 0 {
+                            break;
+                        }
+                    }
+                },
+                None => { 
+                    best_range_tuple = Some(range);
+                }
+            }
         }
+        best_range_tuple.map(|range| {
+            if range.1.end - range.1.start == count {
+                self.empty_ranges.remove(range.0);
+            } else {
+                self.empty_ranges[range.0].start += count;
+            }
+            DualHandle {
+                cpu: d3d12::D3D12_CPU_DESCRIPTOR_HANDLE { ptr: self.start.cpu.ptr + (self.handle_size * range.1.start) as usize },
+                gpu: d3d12::D3D12_GPU_DESCRIPTOR_HANDLE { ptr: self.start.gpu.ptr + (self.handle_size * range.1.start) as u64 },
+                size: count,
+            }
+        })
+    }
+
+    /// Free handles previously given out by this `DescriptorHeapSlice`.  Do not use this with handles not given out by this `DescriptorHeapSlice`.
+    pub(crate) fn free_handles(&mut self, handle: DualHandle) {
+        let start = handle.gpu.ptr - self.start.gpu.ptr / self.handle_size;
+        let handle_range = start..start + handle.size as u64;
+        let mut freed = false;
+        for i in 0..self.empty_ranges.len() {
+            // Handle is immediately to the left of an existing empty range.
+            if handle_range.end == self.empty_ranges[i].start {
+                // Extend this range
+                self.empty_ranges[i].start -= handle_range.end - handle_range.start;
+
+                // Merge this into an adjacent range to the left if necessary.
+                if i > 0 {
+                    if self.empty_ranges[i - 1].end == self.empty_ranges[i].start {
+                        self.empty_ranges[i - 1].end += self.empty_ranges[i].end - self.empty_ranges[i].start;
```

Comment:
similarly, this could just be `self.empty_ranges[i - 1].end = self.empty_ranges[i].end;`